File: src/portfolio_manager/repositories/order_execution_repository.py
"""Order execution repository."""

import json
from typing import Optional
from uuid import uuid4

from portfolio_manager.core.time import now_kst
from portfolio_manager.models.order_execution import OrderExecutionRecord
from portfolio_manager.services.database import OrderExecutionModel


class OrderExecutionRepository:
    """Repository for managing order execution records."""

    def create(
        self,
        ticker: str,
        side: str,
        quantity: int,
        currency: str,
        status: str,
        message: str,
        exchange: Optional[str] = None,
        raw_response: Optional[dict] = None,
    ) -> OrderExecutionRecord:
        """Create an order execution record."""
        now = now_kst()
        row = OrderExecutionModel.create(
            id=uuid4(),
            ticker=ticker,
            side=side,
            quantity=quantity,
            currency=currency,
            exchange=exchange,
            status=status,
            message=message,
            raw_response=json.dumps(raw_response) if raw_response is not None else None,
            created_at=now,
        )
        return self._to_domain(row)

    def list_recent(self, limit: int = 20) -> list[OrderExecutionRecord]:
        """List recent order executions ordered by created_at descending."""
        return [
            self._to_domain(row)
            for row in OrderExecutionModel.select()
            .order_by(OrderExecutionModel.created_at.desc())
            .limit(limit)
        ]

    @staticmethod
    def _to_domain(row: OrderExecutionModel) -> OrderExecutionRecord:
        raw = row.raw_response
        if isinstance(raw, str):
            raw = json.loads(raw)

        return OrderExecutionRecord(
            id=row.id,
            ticker=row.ticker,
            side=row.side,
            quantity=row.quantity,
            currency=row.currency,
            exchange=row.exchange,
            status=row.status,
            message=row.message,
            raw_response=raw,
            created_at=row.created_at,
        )

File: internal/repositories/order_execution_repository.go
package repositories

import (
	"context"
	"database/sql"
	"encoding/json"

	"github.com/kadragon/portfolio-manager/internal/db/sqlc"
	"github.com/kadragon/portfolio-manager/internal/ktime"
	"github.com/kadragon/portfolio-manager/internal/models"
	"github.com/kadragon/portfolio-manager/internal/uuidx"
)

// OrderExecutionRepository persists order execution records.
type OrderExecutionRepository struct {
	q *sqlc.Queries
}

// NewOrderExecutionRepository builds the repository.
func NewOrderExecutionRepository(q *sqlc.Queries) *OrderExecutionRepository {
	return &OrderExecutionRepository{q: q}
}

// Create inserts a new order execution record.
func (r *OrderExecutionRepository) Create(
	ctx context.Context,
	ticker, side string,
	quantity int,
	currency, status, message string,
	exchange string,
	rawResponse map[string]any,
) (models.OrderExecutionRecord, error) {
	var exch sql.NullString
	if exchange != "" {
		exch = sql.NullString{String: exchange, Valid: true}
	}

	var rawJSON sql.NullString
	if rawResponse != nil {
		b, err := json.Marshal(rawResponse)
		if err != nil {
			return models.OrderExecutionRecord{}, err
		}
		rawJSON = sql.NullString{String: string(b), Valid: true}
	}

	row, err := r.q.CreateOrderExecution(ctx, sqlc.CreateOrderExecutionParams{
		ID:          uuidx.New(),
		Ticker:      ticker,
		Side:        side,
		Quantity:    int64(quantity),
		Currency:    currency,
		Exchange:    exch,
		Status:      status,
		Message:     message,
		RawResponse: rawJSON,
		CreatedAt:   ktime.Now(),
	})
	if err != nil {
		return models.OrderExecutionRecord{}, err
	}
	return toDomainOrderExecution(row), nil
}

// ListRecent returns the most recent executions ordered by created_at DESC.
func (r *OrderExecutionRepository) ListRecent(ctx context.Context, limit int) ([]models.OrderExecutionRecord, error) {
	rows, err := r.q.ListRecentOrderExecutions(ctx, int64(limit))
	if err != nil {
		return nil, err
	}
	result := make([]models.OrderExecutionRecord, 0, len(rows))
	for _, row := range rows {
		result = append(result, toDomainOrderExecution(row))
	}
	return result, nil
}

func toDomainOrderExecution(row sqlc.OrderExecution) models.OrderExecutionRecord {
	var raw map[string]any
	if row.RawResponse.Valid && row.RawResponse.String != "" {
		_ = json.Unmarshal([]byte(row.RawResponse.String), &raw)
	}

	exchange := ""
	if row.Exchange.Valid {
		exchange = row.Exchange.String
	}

	return models.OrderExecutionRecord{
		ID:          row.ID,
		Ticker:      row.Ticker,
		Side:        row.Side,
		Quantity:    int(row.Quantity),
		Currency:    row.Currency,
		Exchange:    exchange,
		Status:      row.Status,
		Message:     row.Message,
		RawResponse: raw,
		CreatedAt:   row.CreatedAt,
	}
}
